blib/man3/KiokuDB.3
.\" Automatically generated by Pod::Man 2.22 (Pod::Simple 3.10)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.ie \nF \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
..
.    nr % 0
.    rr F
.\}
.el \{\
.    de IX
..
.\}
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "KiokuDB 3"
.TH KiokuDB 3 "2010-03-21" "perl v5.10.1" "User Contributed Perl Documentation"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
KiokuDB \- Object Graph storage engine
.SH "TUTORIAL"
.IX Header "TUTORIAL"
If you're new to KiokuDB check out KiokuDB::Tutorial.
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
.Vb 1
\&    use KiokuDB;
\&
\&    # use a DSN
\&    my $d = KiokuDB\->connect( $dsn, %args );
\&
\&    # or manually instantiate a backend
\&    my $d = KiokuDB\->new(
\&        backend => KiokuDB::Backend::Files\->new(
\&            dir        => "/tmp/foo",
\&            serializer => "yaml",
\&        ),
\&    );
\&
\&
\&    # create a scope object
\&    my $s = $d\->new_scope;
\&
\&
\&    # takes a snapshot of $some_object
\&    my $uuid = $d\->store($some_object);
\&
\&    # or with a custom ID:
\&    $d\->store( $id => $some_object ); # $id can be any string
\&
\&
\&    # retrieve by ID
\&    my $some_object = $d\->lookup($uuid);
\&
\&
\&
\&    # some backends (like DBI) support simple searchs
\&    $d\->search({ name => "foo" });
\&
\&
\&    # others use GIN queries (DBI supports both)
\&    $d\->search($gin_query);
.Ve
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
KiokuDB is a Moose based frontend to various data stores, somewhere in
between Tangram and Pixie.
.PP
Its purpose is to provide persistence for \*(L"regular\*(R" objects with as little
effort as possible, without sacrificing control over how persistence is
actually done, especially for harder to serialize objects.
.PP
KiokuDB is also non-invasive: it does not use ties, \f(CW\*(C`AUTOLOAD\*(C'\fR, proxy
objects, \f(CW\*(C`sv_magic\*(C'\fR or any other type of trickery.
.PP
Many features important for proper Perl space semantics are supported,
including shared data, circular structures, weak references, tied structures,
etc.
.PP
KiokuDB is meant to solve two related persistence problems:
.IP "Transparent persistence" 4
.IX Item "Transparent persistence"
Store arbitrary objects without changing their class definitions or worrying
about schema details, and without needing to conform to the limitations of
a relational model.
.IP "Interoperability" 4
.IX Item "Interoperability"
Persisting arbitrary objects in a way that is compatible with existing
data/code (for example interoprating with another app using CouchDB with \s-1JSPON\s0
semantics).
.SH "FUNDAMENTAL CONCEPTS"
.IX Header "FUNDAMENTAL CONCEPTS"
In order to use any persistence framework it is important to understand what it
does and how it does it.
.PP
Systems like Tangram or DBIx::Class generally require explicit meta data
and use a schema, which makes them fairly predictable.
.PP
When using transparent systems like KiokuDB or Pixie it is more important
to understand what's going on behind the scenes in order to avoid surprises and
limitations.
.PP
An architectural overview is available on the website:
<http://www.iinteractive.com/kiokudb/arch.html>
.PP
The process is explained here and in the various component documentation in
more detail.
.SS "Collapsing"
.IX Subsection "Collapsing"
When an object is stored using KiokuDB it's collapsed into an
KiokDB::Entry.
.PP
An entry is a simplified representation of the object, allowing the data to be
saved in formats as simple as \s-1JSON\s0.
.PP
References to other objects are converted to symbolic references in the entry,
so objects can be saved independently of each other.
.PP
The entries are given to the KiokuDB::Backend for actual storage.
.PP
Collapsing is explained in detail in KiokuDB::Collapser. The way an entry is
created varies with the object's class.
.SS "Linking"
.IX Subsection "Linking"
When objects are loaded, entries are retrieved from the backend using their
UIDs.
.PP
When a \s-1UID\s0 is already loaded (in the live object set of a KiokuDB instance,
see KiokuDB::LiveObjects) the live object is used. This way references to
shared objects are shared in memory regardless of the order the objects were
stored or loaded.
.PP
This process is explained in detail in KiokuDB::Linker.
.SH "ROOT SET MEMBERSHIP"
.IX Header "ROOT SET MEMBERSHIP"
Any object that is passed to \f(CW\*(C`store\*(C'\fR or \f(CW\*(C`insert\*(C'\fR directly is implicitly
considered a member of the root set.
.PP
This flag implies that the object is an identified resource and should not be
garbage collected with any of the proposed garbage collection schemes.
.PP
The root flag may be modified explicitly:
.PP
.Vb 1
\&    $kiokudb\->set_root(@objects); # or unset_root
\&
\&    $kiokudb\->update(@objects);
.Ve
.PP
Lastly, root set membership may also be specified explicitly by the typemap.
.PP
A root set member must be explicitly removed using \f(CW\*(C`delete\*(C'\fR or by
removing it from the root set. Only non-members of the root set will be
purged with any garbage collection scheme.
.SH "TRANSACTIONS"
.IX Header "TRANSACTIONS"
On supporting backends the \f(CW\*(C`txn_do\*(C'\fR method will execute a block and commit the
transaction at its end.
.PP
Nesting of \f(CW\*(C`txn_do\*(C'\fR blocks is always supported, though rolling back a nested
transaction may produce different results on different backends.
.PP
If the backend does not support transactions \f(CW\*(C`txn_do\*(C'\fR simply executes the code
block normally.
.SH "CONCURRENCY"
.IX Header "CONCURRENCY"
Most transactional backends are also concurrent.
.PP
KiokuDB::Backend::BDB and KiokuDB::Backend::CouchDB default to
serializable transaction isolation and do not suffer from deadlocks, but
serialization errors may occur, aborting the transaction (in which case the
transaction should be tried again).
.PP
KiokuDB::Backend::Files provides good concurrency support but will only
detect deadlocks on platforms which return \f(CW\*(C`EDEADLK\*(C'\fR from \f(CW\*(C`flock\*(C'\fR.
Directory::Transactional may provide alternative mechanisms in the future.
.PP
Concurrency support in KiokuDB::Backend::DBI depends on the database. SQLite
defaults to serializable transaction isolation out of the box, wheras MySQL and
PostgreSQL default to read committed.
.PP
Depending on your application read committed isolation may be sufficient, but
due to the graph structure nature of the data repeatable reads or serializable
level isolation is highly reccomended. Read committed isolation generally works
well when each row in the database is more or less independent of others, and
various constraints ensure integrity. Unfortunately this is not the case with
the graph layout.
.PP
To enable stronger isolation guarantees see
\&\*(L"Transactions\*(R" in KiokuDB::Backend::DBI for per-database pointers.
.SH "ATTRIBUTES"
.IX Header "ATTRIBUTES"
KiokuDB uses a number of delegates which do the actual work.
.PP
Of these only \f(CW\*(C`backend\*(C'\fR is required, the rest have default definitions.
.PP
Additional attributes that are not commonly used are listed in \*(L"\s-1INTERNAL\s0
\&\s-1ATTRIBUTES\s0\*(R".
.IP "backend" 4
.IX Item "backend"
This attribute is required.
.Sp
This must be an object that does KiokuDB::Backend.
.Sp
The backend handles storage and retrieval of entries.
.IP "typemap" 4
.IX Item "typemap"
This is an instance KiokuDB::TypeMap.
.Sp
The typemap contains entries which control how KiokuDB::Collapser and
KiokuDB::Linker handle different types of objects.
.IP "allow_classes" 4
.IX Item "allow_classes"
An array references of extra classes to allow.
.Sp
Objects blessed into these classes will be collapsed using
KiokuDB::TypeMap::Entry:Naive.
.IP "allow_bases" 4
.IX Item "allow_bases"
An array references of extra base classes to allow.
.Sp
Objects derived from these classes will be collapsed using
KiokuDB::TypeMap::Entry:Naive.
.IP "allow_class_builders" 4
.IX Item "allow_class_builders"
If true adds KiokuDB::TypeMap::ClassBuilders to the merged typemap.
.Sp
It's possible to provide a hash reference of options to give to
\&\*(L"new\*(R" in KiokuDB::TypeMap::ClassBuilders.
.IP "check_class_versions" 4
.IX Item "check_class_versions"
Controls whether or not the class versions of objects are checked on load.
.Sp
Defaults to true.
.IP "class_version_table" 4
.IX Item "class_version_table"
A table of classes and versions that is passed to the default typemap entry for
Moose/Class::MOP objects.
.Sp
When a class version has changed between the time that an object was stored and
the time it's being retrieved, the data must be converted.
.Sp
See KiokuDB::TypeMap::Entry::MOP for more details.
.SH "METHODS"
.IX Header "METHODS"
.ie n .IP "connect $dsn, %args" 4
.el .IP "connect \f(CW$dsn\fR, \f(CW%args\fR" 4
.IX Item "connect $dsn, %args"
\&\s-1DWIM\s0 wrapper for \f(CW\*(C`new\*(C'\fR.
.Sp
\&\f(CW$dsn\fR represents some sort of backend (much like \s-1DBI\s0 dsns map to DBDs).
.Sp
An example \s-1DSN\s0 is:
.Sp
.Vb 1
\&    my $dir = KiokuDB\->connect("bdb:dir=path/to/data/");
.Ve
.Sp
The backend moniker name is extracted by splitting on the colon. The rest of
the string is passed to \f(CW\*(C`new_from_dsn\*(C'\fR, which is documented in more detail in
KiokuDB::Backend.
.Sp
Typically \s-1DSN\s0 arguments are separated by \f(CW\*(C`;\*(C'\fR, with \f(CW\*(C`=\*(C'\fR separating keys and
values. Arguments with no value are assumed to denote boolean truth (e.g.
\&\f(CW\*(C`jspon:dir=foo;pretty\*(C'\fR means \f(CW\*(C`dir => "foo", pretty => 1\*(C'\fR). However, a
backend may override the default parsing, so this is not guaranteed.
.Sp
Extra arguments are passed both to the backend constructor, and the \f(CW\*(C`KiokuDB\*(C'\fR
constructor.
.Sp
Note that if you need a typemap you still need to pass it in:
.Sp
.Vb 1
\&    KiokuDB\->connect( $dsn, typemap => $typemap );
.Ve
.Sp
The \s-1DSN\s0 can also be a valid \s-1JSON\s0 string taking one of the following forms:
.Sp
.Vb 1
\&    dsn => \*(Aq["dbi:SQLite:foo",{"schema":"MyApp::DB"}]\*(Aq
\&
\&    dsn => \*(Aq{"dsn":"dbi:SQLite:foo","schema":"MyApp::DB"}\*(Aq
.Ve
.Sp
This allows more complicated arguments to be specified accurately, or arbitrary
options to be specified when the backend has nonstandard \s-1DSN\s0 parsing (for
instance KiokuDB::Backend::DBI simply passes the string to \s-1DBI\s0, so this
is necessary in order to specify options on the command line).
.ie n .IP "configure $config_file, %args" 4

lib/KiokuDB.pm
#!/usr/bin/perl

package KiokuDB;
use Moose;

our $VERSION = "0.14";

use constant SERIAL_IDS => not not our $SERIAL_IDS;

use KiokuDB::Backend;
use KiokuDB::Collapser;
use KiokuDB::Linker;
use KiokuDB::LiveObjects;
use KiokuDB::TypeMap;
use KiokuDB::TypeMap::Shadow;
use KiokuDB::TypeMap::Resolver;

use Hash::Util::FieldHash::Compat qw(idhash);
use Carp qw(croak);

use namespace::clean -except => [qw(meta SERIAL_IDS)];

sub connect {
    my ( $class, $dsn, @args ) = @_;

    if ( -d $dsn ) {
        return $class->configure($dsn, @args);
    } else {
        require KiokuDB::Util;
        return $class->new( backend => KiokuDB::Util::dsn_to_backend($dsn, @args), @args );
    }
}

sub configure {
    my ( $class, $base, @args ) = @_;

    require Path::Class;
    $base = Path::Class::dir($base) unless blessed $base;

    require KiokuDB::Util;
    my $config = KiokuDB::Util::load_config($base);

    my $backend = KiokuDB::Util::config_to_backend( $config, base => $base, @args );

    # FIXME gin extractor, typemap, etc
    $class->new( %$config, @args, backend => $backend );
}

has typemap => (
    does => "KiokuDB::Role::TypeMap",
    is   => "ro",
    lazy_build => 1,
);

sub _build_typemap {
    KiokuDB::TypeMap->new;
}

has merged_typemap => (
    does => "KiokuDB::Role::TypeMap",
    is   => "ro",
    lazy_build => 1,
);

sub _find_default_typemap {
    my $self = shift;

    my $b = $self->backend;

    if ( $b->can("default_typemap") ) {
        return $b->default_typemap;
    } elsif( $b->can("serializer") and $b->serializer->can("default_typemap") ) {
        return $b->serializer->default_typemap;
    }

    return;
}

sub _build_merged_typemap {
    my $self = shift;

    if ( my $default_typemap = $self->_find_default_typemap ) {
        return KiokuDB::TypeMap::Shadow->new(
            typemaps => [
                $self->typemap,
                $default_typemap,
            ],
        );
    } else {
        return $self->typemap;
    }
}

has typemap_resolver => (
    isa => "KiokuDB::TypeMap::Resolver",
    is  => "ro",
    lazy_build => 1,
);

sub _build_typemap_resolver {
    my $self = shift;

    KiokuDB::TypeMap::Resolver->new(
        typemap => $self->merged_typemap,
    );
}

has live_objects => (
    isa => "KiokuDB::LiveObjects",
    is  => "ro",
    lazy => 1,
    builder => "_build_live_objects", # lazy_build => 1 sets clearer
    handles => {
        clear_live_objects => "clear",
        new_scope          => "new_scope",
        object_to_id       => "object_to_id",
        "objects_to_ids"   => "objects_to_ids"
    },
);

sub _build_live_objects { KiokuDB::LiveObjects->new }

has collapser => (
    isa => "KiokuDB::Collapser",
    is  => "ro",
    lazy_build => 1,
);

sub _build_collapser {
    my $self = shift;

    KiokuDB::Collapser->new(
        live_objects => $self->live_objects,
        typemap_resolver => $self->typemap_resolver,
    );
}

has backend => (
    does => "KiokuDB::Backend",
    is   => "ro",
    required => 1,
    coerce   => 1,
    handles  => [qw(exists)],
);

has linker => (
    isa => "KiokuDB::Linker",
    is  => "ro",
    lazy_build => 1,
);

sub _build_linker {
    my $self = shift;

    KiokuDB::Linker->new(
        backend => $self->backend,
        live_objects => $self->live_objects,
        typemap_resolver => $self->typemap_resolver,
    );
}

sub lookup {
    my ( $self, @ids ) = @_;

    my $linker = $self->linker;

    my ( $e, @objects );

    eval {
        local $@;
        eval { @objects = $linker->get_or_load_objects(@ids) };
        $e = $@;
    };

    if ( $e ) {
        if ( ref($e) and $e->{missing} ) {
            return;
        }

        die $e;
    }

    if ( @ids == 1 ) {
        return $objects[0];
    } else {
        return @objects;
    }
}

sub search {
    my ( $self, @args ) = @_;

    if ( @args == 1 && ref $args[0] eq 'HASH' ) {
        return $self->simple_search(@args);
    } else {
        return $self->backend_search(@args);
    }
}

sub _load_entry_stream {
    my ( $self, $entries ) = @_;

    my $linker = $self->linker;

    return $entries->filter(sub {[ $linker->load_entries(@$_) ]});
}

sub simple_search {
    my ( $self, @args ) = @_;

    my $b = $self->backend;

    my $entries = $b->simple_search( @args, live_objects => $self->live_objects );

    my $objects = $self->_load_entry_stream($entries);

    return $b->simple_search_filter($objects, @args);
}

sub backend_search {
    my ( $self, @args ) = @_;

    my $b = $self->backend;

    my $entries = $b->search( @args, live_objects => $self->live_objects );

    my $objects = $self->_load_entry_stream($entries);

    return $b->search_filter($objects, @args);
}

sub root_set {
    my ( $self ) = @_;

    my $entries = $self->backend->root_entries;

    my $linker = $self->linker;

    $entries->filter(sub {[ $linker->load_entries(@$_) ]});
}

sub all_objects {
    my ( $self ) = @_;

    my $entries = $self->backend->all_entries;

    my $linker = $self->linker;

    $entries->filter(sub {[ $linker->load_entries(@$_) ]});
}

sub grep {
    my ( $self, $filter ) = @_;

    my $stream = $self->root_set;

    $stream->filter(sub { [ grep { $filter->($_) } @$_ ] });
}

sub scan {
    my ( $self, $filter ) = @_;

    my $stream = $self->root_set;

    while ( my $items = $stream->next ) {
        foreach my $item ( @$items ) {
            $item->$filter();
        }
    }
}

sub _parse_args {
    my ( $self, @args ) = @_;

    my ( %ids, @ret );

    while ( @args ) {
        my $next = shift @args;

        unless ( ref $next ) {
            my $obj = shift @args;

            $ids{$next} = $obj;

            push @ret, $obj;
        } else {
            push @ret, $next;
        }
    }

    return ( \%ids, @ret );
}

sub _register {
    my ( $self, @args ) = @_;

    my ( $ids, @objs ) = $self->_parse_args(@args);

    if ( scalar keys %$ids ) {
        $self->live_objects->insert(%$ids);
    }

    return @objs;
}

sub store {
    my ( $self, @args ) = @_;

    my @objects = $self->_register(@args);

    $self->store_objects( root_set => 1, objects => \@objects );
}

sub insert {
    my ( $self, @args ) = @_;

    my @objects = $self->_register(@args);

    idhash my %entries;

    @entries{@objects} = $self->live_objects->objects_to_entries(@objects);

    # FIXME make optional?
    if ( my @in_storage = grep { $entries{$_} } @objects ) {
        croak "Objects already in database: @in_storage";
    }

    $self->store_objects( root_set => 1, only_new => 1, objects => \@objects );

    # return IDs only for unknown objects
    if ( defined wantarray ) {
        return $self->live_objects->objects_to_ids(@objects);
    }
}

sub update {
    my ( $self, @args ) = @_;

    my @objects = $self->_register(@args);

    my $l = $self->live_objects;

    croak "Object not in storage"
        if grep { not defined } $l->objects_to_entries(@objects);

    $self->store_objects( shallow => 1, only_known => 1, objects => \@objects );
}

sub store_objects {
    my ( $self, %args ) = @_;

    my $objects = $args{objects};

    my ( $entries, @ids ) = $self->collapser->collapse(%args);

    if ( $args{root_set} ) {
        $_->root(1) for grep { defined } @{$entries}{@ids};
    }

    $self->backend->insert(values %$entries);

    $self->live_objects->update_entries(values %$entries);

    if ( @$objects == 1 ) {
        return $ids[0];
    } else {
        return @ids;
    }
}

sub delete {
    my ( $self, @ids_or_objects ) = @_;

    my $l = $self->live_objects;

    my ( @ids, @objects );

    push @{ ref($_) ? \@objects : \@ids }, $_ for @ids_or_objects;

    my @entries;

    push @entries, $l->objects_to_entries(@objects) if @objects;

    for ( @entries ) {
        croak "Object not in storage" unless defined;
    }

    @entries = map { $_->deletion_entry } @entries;

    # FIXME ideally if ID is pointing at a live object we should use its entry
    #push @entries, $l->ids_to_entries(@ids) if @ids;
    my @ids_or_entries = ( @entries, @ids );

    $self->backend->delete(@ids_or_entries);

    $l->update_entries(@entries);
}

sub txn_do {
    my ( $self, $code, %args ) = @_;

    my $backend = $self->backend;

    if ( $backend->can("txn_do") ) {
        my $scope = $self->live_objects->new_txn;

        my $rollback = $args{rollback};
        $args{rollback} = sub { $scope->rollback; $rollback && $rollback->() };

        $backend->txn_do( $code, %args );
    } else {
        return $code->();
    }
}

__PACKAGE__->meta->make_immutable;

__PACKAGE__

__END__

=pod

=head1 NAME

KiokuDB - Object Graph storage engine

=head1 SYNOPSIS

    use KiokuDB;

    # use a DSN
    my $d = KiokuDB->connect( $dsn, %args );

    # or manually instantiate a backend
    my $d = KiokuDB->new(
        backend => KiokuDB::Backend::Files->new(
            dir        => "/tmp/foo",
            serializer => "yaml",
        ),
    );


    # takes a snapshot of $some_object
    my $uuid = $d->store($some_object);

    # or with a custom ID:
    $d->store( $id => $some_object ); # $id can be any string


    # retrieve by ID
    my $some_object = $d->lookup($uuid);



    # some backends (like DBI) support simple searchs
    $d->search({ name => "foo" });


    # others use GIN queries (DBI supports both)
    $d->search($gin_query);

=head1 ALPHA WARNING

L<KiokuDB> is still in its infancy. No fundamental changes are expected, but
nevertheless backwards compatibility is not yet guaranteed.

=head1 DESCRIPTION

Kioku is a Moose based frontend to various databases, somewhere in between
L<Tangram> and L<Pixie>. It builds on L<Class::MOP>'s solid foundation.

Its purpose is to provide persistence for "regular" objects with as little
effort as possible, without sacrificing control over how persistence is
actually done.

Kioku is also non-invasive: it does not use ties, AUTOLOAD, proxy objects,
C<sv_magic> or any other type of trickery to get its job done, to avoid
unwanted surprises.

Many features important for proper Perl space semantics are supported,
including shared data, circular structures, weak references, tied structures,
etc.

L<KiokuDB> is meant to solve two related persistence problems:

=over 4

=item Transparent persistence

Store arbitrary objects without changing their class definitions or worrying
about schema details, and without needing to conform to the limitations of
a relational model.

=item Interoperability

Persisting arbitrary objects in a way that is compatible with existing
data/code (for example interoprating with another app using CouchDB with JSPON
semantics).

=back

=head1 FUNDAMENTAL CONCEPTS

In order to use any persistence framework it is important to understand what it
does and how it does it.

Systems like L<Tangram> or L<DBIx::Class> generally require explicit meta data
and use a schema, which makes them fairly predictable.

When using transparent systems like L<KiokuDB> or L<Pixie> it is more important
to understand what's going on behind the scenes in order to avoid surprises and
limitations.

=head2 Collapsing

When an object is stored using L<KiokuDB> it's collapsed into an
L<KiokDB::Entry|Entry>.

An entry is a simplified representation of the object, allowing the data to be
saved in formats as simple as JSON.

References to other objects are converted to symbolic references in the entry,
so objects can be saved independently of each other.

The entries are given to the L<KiokuDB::Backend|Backend> for actual storage.

Collapsing is explained in detail in L<KiokuDB::Collapser>. The way an entry is
created varies with the object's class.

=head2 Linking

When objects are loaded, entries are retrieved from the backend using their
UIDs.

When a UID is already loaded (in the live object set of a L<KiokuDB> instance,
see L<KiokuDB::LiveObjects>) the live object is used. This way references to
shared objects are shared in memory regardless of the order the objects were
stored or loaded.

This process is explained in detail in L<KiokuDB::Linker>.

=head1 ATTRIBUTES

L<KiokuDB> uses a number of delegates which do the actual work.

Of these only C<backend> is required, the rest have default definitions.

Additional attributes that are not commonly used are listed in L</"INTERNAL
ATTRIBUTES">.

=over 4

=item backend

This attribute is required.

This must be an object that does L<KiokuDB::Backend>.

The backend handles storage and retrieval of entries.

=item typemap

This is an instance L<KiokuDB::TypeMap>.

The typemap contains entries which control how L<KiokuDB::Collapser> and
L<KiokuDB::Linker> handle different types of objects.

=back

=head1 METHODS

=over 4

=item connect $dsn, %args

DWIM wrapper for C<new>.

C<$dsn> represents some sort of backend (much like L<DBI> dsns map to DBDs).

An example DSN is:

    my $dir = KiokuDB->connect("bdb:dir=path/to/data/");

The backend moniker name is extracted by splitting on the colon. The rest of
the string is passed tp C<new_from_dsn>, which is documented in more detail in
L<KiokuDB::Backend>.

Typically DSN arguments are separated by C<;>, with C<=> separating keys and
values. Arguments with no value are assumed to denote boolean truth (e.g.
C<jspon:dir=foo;pretty> means c<dir => "foo", pretty => 1>).

Extra arguments are passed both to the backend constructor, and the C<KiokuDB>
constructor.

Note that if you need a typemap you still need to pass it in:

    KiokuDB->connect( $dsn, typemap => $typemap );

=item configure $config_file, %args

TODO

=item new %args

Creates a new directory object.

See L</ATTRIBUTES>

=item new_scope

Creates a new object scope. Handled by C<live_objects>.

The object scope artificially bumps up the reference count of objects to ensure
that they live at least as long as the scope does.

This ensures that weak references aren't deleted prematurely, and the object
graph doesn't get corrupted without needing to create circular structures and
cleaning up leaks manually.

=item lookup @ids

Fetches the objects for the specified IDs from the live object set or from
storage.

=item store @objects

Recursively collapses C<@objects> and inserts or updates the entries.

This performs a full update of every reachable object from C<@objects>,
snapshotting everything.

=item update @objects

Performs a shallow update of @objects.

It is an error to update an object not in the database.

=item insert @objects

Inserts objects to the database.

It is an error to insert objects that are already in the database, all elements
of C<@objects> must be new.

C<@objects> will be collapsed recursively, but the collapsing stops at known
objects, which will not be updated.

=item delete @objects_or_ids

Deletes the specified objects from the store.

Note that this can cause lookup errors if the object you are deleting is
referred to by another object, because that link will be broken.

=item txn_do $code, %args

Executes $code within the scope of a transaction.

This requires that the backend supports transactions
(L<KiokuDB::Backend::Role::TXN>).

Transactions may be nested.

=item search \%proto

=item search @args

Searching requires a backend that supports querying.

The C<\%proto> form is currently unspecified but in the future should provide a
simple but consistent way of looking objects by attributes.

The second form is backend specific querying, for instance
L<Search::GIN::Query> objects passed to L<KiokuDB::Backend::BDB::GIN> or
the generic GIN backend wrapper L<KiokuDB::GIN>.

=item root_set

Returns a L<Data::Stream::Bulk> of all the root objects in the database.

=item all_objects

Returns a L<Data::Stream::Bulk> of all the objects in the database.

=item grep $filter

Returns a L<Data::Stream::Bulk> of the objects in C<root_set> filtered by
C<$filter>.

=item scan $callback

Iterates the root set calling C<$callback> for each object.

=back

=head1 GLOBALS

=over 4

=item C<$SERIAL_IDS>

If set at compile time, the default UUID generation role will use serial IDs,
instead of UUIDs.

This is useful for testing, since the same IDs will be issued each run, but is
utterly broken in the face of concurrency.

=back

=head1 INTERNAL ATTRIBUTES

These attributes are documented for completeness and should typically not be
needed.

=over 4

=item collapser

L<KiokuDB::Collapser>

The collapser prepares objects for storage, by creating L<KiokDB::Entry>
objects to pass to the backend.

=item linker

L<KiokuDB::Linker>

The linker links entries into functioning instances, loading necessary
dependencies from the backend.

=item live_objects

L<KiokuDB::LiveObjects>

The live object set keeps track of objects and entries for the linker and the
resolver.

It also creates scope objects that help ensure objects don't garbage collect
too early (L<KiokuDB::LiveObjects/new_scope>, L<KiokuDB::LiveObjects::Scope>),
and transaction scope objects used by C<txn_do>
(L<KiokuDB::LiveObjects::TXNScope>).

=item typemap_resolver

An instance of L<KiokuDB::TypeMap::Resolver>. Handles actual lookup and
compilation of typemap entries, using the user typemap.

=back

=head1 SEE ALSO

=head2 Prior Art on the CPAN

=over 4

=item L<Pixie>

=item L<DBM::Deep>

=item L<OOPS>

=item L<Tangram>

=item L<DBIx::Class>

Polymorphic retrieval is possible with L<DBIx::Class::DynamicSubclass>

=item L<Fey::ORM>

=item L<MooseX::Storage>

=back

=head2 Links

L<http://en.wikipedia.org/wiki/Navigational_database>

L<http://www.postgresql.org/docs/8.2/static/gin-extensibility.html>

L<http://www.zope.org/Documentation/How-To/ZCatalogTutorial>

L<http://www.couchdb.org>

L<http://sitepen.com/labs/persevere.php>, L<http://code.google.com/p/persevere-framework/>

L<http://www.jspon.org/>

L<http://www.zope.org/Products/StandaloneZODB>

L<http://en.wikipedia.org/wiki/Language_Integrated_Query>

L<http://en.wikipedia.org/wiki/OQL>

=head1 VERSION CONTROL

KiokuDB is maintained using Git. Information about the repository is available
on L<http://www.iinteractive.com/kiokudb/>

=head1 AUTHOR

Yuval Kogman E<lt>nothingmuch@woobling.orgE<gt>

=head1 COPYRIGHT

    Copyright (c) 2008 Yuval Kogman, Infinity Interactive. All rights
    reserved This program is free software; you can redistribute
    it and/or modify it under the same terms as Perl itself.

=cut
